src/core/experimental/refcount.d

```diff
@@ -0,0 +1,548 @@
+    }
+
+    /**
+    Copy constructs a $(B const shared __RefCount) from a $(B const shared)
+    reference, `rhs`. This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope const shared ref typeof(this) rhs) const shared
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Copy constructs a $(B const __RefCount) from an `immutable` reference, `rhs`.
+    This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope immutable ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Copy constructs a $(B const shared __RefCount) from an $(B immutable shared)
+    reference, `rhs`. This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope immutable shared ref typeof(this) rhs) const shared
+    {
+        mixin(copyCtorIncRef);
+    }
+
+
+    /**
+    Copy construct an $(B immutable __RefCount) from an `immutable` reference, `rhs`.
+    This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope immutable ref typeof(this) rhs) immutable
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Assign a `__RefCount` object into this. This will decrement the old reference
+    count before assigning the new one. If the old reference was the last one,
+    this will trigger the deallocation of the old ref. This increases the
+    reference count of `rhs`.
+
+    Params:
+         rhs = the `__RefCount` object to be assigned.
+
+    Returns:
+         A reference to `this`.
+
+    Complexity:
+         $(BIGOH 1).
+    */
+    @nogc nothrow pure @safe scope
+    ref __RefCount opAssign(return scope ref typeof(this) rhs) return
+    {
+        if (rhs.isInitialized() && rc == rhs.rc)
+        {
+            return this;
+        }
+        if (rhs.isInitialized())
+        {
+            rhs.addRef();
+        }
+        if (isInitialized())
+        {
+            delRef();
+        }
+        () @trusted { rc = rhs.rc; }();
+        return this;
+    }
+
+    ///
+    @nogc nothrow pure @safe scope
+    ref shared(__RefCount) opAssign(return scope shared ref typeof(this) rhs) return shared
+    {
+        if (rhs.isInitialized() && rc == rhs.rc)
+        {
+            return this;
+        }
+        if (rhs.isInitialized())
+        {
+            rhs.addRef();
+        }
+        if (isInitialized())
+        {
+            delRef();
+        }
+        () @trusted { rc = rhs.rc; }();
+        return this;
+    }
+
+    /*
+    Increase the reference count. This asserts that `__RefCount` is initialized.
+
+    Returns:
+         This returns a `void*` so the compiler won't optimize away the call
+         to this `const pure` function.
+    */
+    @nogc nothrow pure @safe scope
+    private void* addRef(this Q)() const
+    {
+        assert(isInitialized(), "[__RefCount.addRef] __RefCount is uninitialized");
```

Comment:
Maybe use `.stringof` to get the full name of the RefCount member, s.t. the user knows which one it was?

Reply:
That just prints `module refcount` (because `__RefCount`isn't a template?). Anyway, this is `private` and the stacktrace should be enough, so I'm just removing the bracketed stuff.